File: apps/server/src/cocktail_server/schemas/conversations.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from cocktail_server.schemas.images import GeneratedImageRef
from cocktail_server.schemas.messages import Message


class ConversationDetail(BaseModel):
    """`GET /conversations/{id}` の応答。URL 直開き / リロードで会話を復元するためのスナップショット。"""

    model_config = ConfigDict(strict=True, extra="forbid")

    id: str = Field(min_length=1, max_length=128)
    created_at: datetime
    updated_at: datetime
    messages: list[Message]
    generated_images: list[GeneratedImageRef]


class ConversationSummary(BaseModel):
    """`GET /conversations` の応答要素。左メニューの履歴一覧向けに軽量なメタのみ持つ。

    `title` は会話識別用の短いラベル（最初のユーザテキスト先頭 40 文字、なければ "(新規会話)"）。
    """

    model_config = ConfigDict(strict=True, extra="forbid")

    id: str = Field(min_length=1, max_length=128)
    title: str = Field(min_length=1, max_length=80)
    created_at: datetime
    updated_at: datetime
    message_count: int = Field(ge=0)

File: packages/api-types/src/conversations.ts
import { z } from "zod";

import { GeneratedImageRefSchema } from "./images.js";
import { MessageSchema } from "./messages.js";

/**
 * `apps/server/src/cocktail_server/schemas/conversations.py` の `ConversationDetail` を
 * ミラーする。URL 直開き / リロード時に会話を復元する `GET /conversations/{id}` の応答。
 */
export const ConversationDetailSchema = z
  .object({
    id: z.string().min(1).max(128),
    created_at: z.string(),
    updated_at: z.string(),
    messages: z.array(MessageSchema),
    generated_images: z.array(GeneratedImageRefSchema),
  })
  .strict();

export type ConversationDetail = z.infer<typeof ConversationDetailSchema>;

/**
 * `apps/server/src/cocktail_server/schemas/conversations.py` の `ConversationSummary` を
 * ミラーする。左メニューの履歴一覧向け軽量サマリ。
 */
export const ConversationSummarySchema = z
  .object({
    id: z.string().min(1).max(128),
    title: z.string().min(1).max(80),
    created_at: z.string(),
    updated_at: z.string(),
    message_count: z.number().int().min(0),
  })
  .strict();

export type ConversationSummary = z.infer<typeof ConversationSummarySchema>;
